src/index.js

```diff
@@ -1,390 +1,586 @@
-            )
-          : document.createElement(node.nodeName)
-
-    var attributes = node.attributes
-    if (attributes) {
-      if (attributes.oncreate) {
-        lifecycle.push(function() {
-          attributes.oncreate(element)
-        })
-      }
+  return (nextNode.element = element)
+}
 
-      for (var i = 0; i < node.children.length; i++) {
-        element.appendChild(
-          createElement(
-            (node.children[i] = resolveNode(node.children[i])),
-            isSvg
-          )
-        )
-      }
+var createVNode = function(name, props, children, element, key, type) {
+  return {
+    name: name,
+    props: props,
+    children: children,
+    element: element,
+    key: key,
+    type: type
+  }
+}
 
-      for (var name in attributes) {
-        updateAttribute(element, name, attributes[name], null, isSvg)
-      }
-    }
+var createTextVNode = function(text, element) {
+  return createVNode(text, EMPTY_OBJECT, EMPTY_ARRAY, element, null, TEXT_NODE)
+}
 
-    return element
-  }
+var recycleChild = function(element) {
+  return element.nodeType === 3 // Node.TEXT_NODE
+    ? createTextVNode(element.nodeValue, element)
+    : recycleElement(element)
+}
 
-  function updateElement(element, oldAttributes, attributes, isSvg) {
-    for (var name in clone(oldAttributes, attributes)) {
-      if (
-        attributes[name] !==
-        (name === "value" || name === "checked"
-          ? element[name]
-          : oldAttributes[name])
-      ) {
-        updateAttribute(
-          element,
-          name,
-          attributes[name],
-          oldAttributes[name],
-          isSvg
-        )
-      }
-    }
+var recycleElement = function(element) {
+  return createVNode(
+    element.nodeName.toLowerCase(),
+    EMPTY_OBJECT,
+    map.call(element.childNodes, recycleChild),
+    element,
+    null,
+    RECYCLED_NODE
+  )
+}
 
-    var cb = isRecycling ? attributes.oncreate : attributes.onupdate
-    if (cb) {
-      lifecycle.push(function() {
-        cb(element, oldAttributes)
-      })
-    }
-  }
+var patch = function(container, element, lastNode, nextNode, eventProxy) {
+  var lifecycle = []
 
-  function removeChildren(element, node) {
-    var attributes = node.attributes
-    if (attributes) {
-      for (var i = 0; i < node.children.length; i++) {
-        removeChildren(element.childNodes[i], node.children[i])
-      }
+  patchElement(container, element, lastNode, nextNode, lifecycle, eventProxy)
 
-      if (attributes.ondestroy) {
-        attributes.ondestroy(element)
-      }
-    }
-    return element
-  }
+  while (lifecycle.length > 0) lifecycle.pop()()
 
-  function removeElement(parent, element, node) {
-    function done() {
-      parent.removeChild(removeChildren(element, node))
-    }
+  return nextNode
+}
 
-    var cb = node.attributes && node.attributes.onremove
-    if (cb) {
-      cb(element, done)
-    } else {
-      done()
+export var h = function(name, props) {
```

Comment:
As we are exporting h object here so this source file becomes ES6 module. So why then we are not using all power of ES6. To name a few, using const, Object.assign, spread operator, etc which will make code more readable, smaller (not larger at least) and simple without adding any burden to performance. Making methods will easily become one liners.

Reply:
I don't think this file gets transpiled. Rollup supports specifying UMD, which is how this project is configured, which is probably where the export support comes from.
https://github.com/hyperapp/hyperapp/blob/2b3d85625321038ed2369ddb4ecc320a9e20984e/package.json#L27